java/main/src/main/java/com/saucelabs/saucebindings/VisualSession.java

```diff
@@ -0,0 +1,85 @@
+package com.saucelabs.saucebindings;
+
+import com.saucelabs.saucebindings.options.SauceOptions;
+import com.saucelabs.saucebindings.options.VisualOptions;
+import org.openqa.selenium.InvalidArgumentException;
+import org.openqa.selenium.remote.RemoteWebDriver;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class VisualSession extends SauceSession {
+    private VisualResults visualResults;
+    private final VisualOptions visualOptions;
+
+    public VisualSession(String testName) {
+        this(new VisualOptions(SauceOptions.chrome().setName(testName).build()));
+    }
+
+    public VisualSession(VisualOptions options) {
+        super(options.getSauceOptions());
+        this.visualOptions = options;
+    }
+
+    @Override
+    public RemoteWebDriver start() {
+        this.driver = createRemoteWebDriver(getSauceUrl(), visualOptions.toCapabilities());
+        newVisualTest(getSauceOptions().sauce().getName());
+        return driver;
+    }
+
+    @Override
+    public URL getSauceUrl() {
+        try {
+            return new URL("https://hub.screener.io/wd/hub");
+        } catch (MalformedURLException e) {
+            throw new InvalidArgumentException("Invalid URL", e);
+        }
+    }
+
+    public void newVisualTest(String testName) {
+        validateVisualStatus();
+        driver.executeScript("/*@visual.init*/", testName);
+    }
+
+    public void takeSnapshot(String name) {
+        validateVisualStatus();
+        driver.executeScript("/*@visual.snapshot*/", name);
+    }
+
+    public VisualResults getVisualResults() {
+        if (visualResults == null) {
+            this.visualResults = VisualResults.generate(driver);
+        }
+        return visualResults;
+    }
+
+    @Override
+    public void stop(Boolean passed) {
+        String result = passed ? "passed" : "failed";
```

Comment:
magic strings

Reply:
You know, this is a good point. It comes from our using — https://docs.saucelabs.com/basics/test-config-annotation/test-annotation/#methods which is literally *all magic strings.

When we update the implementation to use the API instead of JS to record pass/fail, then the magic strings are completely useless... Yeah, now I want to deprecate the method in `SauceSession` instead of supporting it in `VisualSession`.